keps/sig-apps/0753-sidecarcontainers.md

```diff
@@ -328,9 +334,82 @@ Older Kubelets should still be able to schedule Pods that have sidecar container
+```yaml
+initContainers:
+  - name: myInit
+containers:
+  - name: myApp
+sidecarContainers:
+  - name: mySidecar
+```
+
+### Primary Container
+The primary container idea was specific to solving the issue of Jobs that don't complete with a sidecar, the suggestion was to have one container marked as the primary so that the Job would get completed when that container has finished. This was dismissed as it was too specific to Jobs whereas the more generic issues of sidecars could be useful in other places.
+```yaml
+kind: Job
+spec:
+  template:
+    spec:
+      containers:
+      - name: worker
+        command: ["do a job"]
+      - name: "sidecar"
+        command: ["help"]
+  backoffLimit: 4
+  primaryContainer: worker
+```
 
-Another alternative would be to change the Job Spec to have a `primaryContainer` field to tell it which containers are important. However I feel this is perhaps too specific to job when this Sidecar concept could be useful in other scenarios.
+### Sidecar: true
+```yaml
+containers:
+  - name: myApp
+  - name: mySidecar
+    sidecar: true
+```
+A boolean flag of `sidecar: true` could be used to indicate which pods are sidecars, this was dismissed as it was considered too specific and potentially other types of container lifecycle may want to be added in the future.
 
-A boolean flag of `sidecar: true` could be used to indicate which pods are sidecars, however this prevents additional ContainerLifecycles from being added in the future.
+### terminationFatalToPod: true
+This suggestion was to have the ability to mark certain containers as critical to the pod, if they exited it would cause the other containers to exit. While this helped solve things like Jobs it didn't solve the wider issue of ordering startup and shutdown.
+
+```yaml
+containers:
+  - name: myApp
+    terminationFatalToPod: true
+  - name: mySidecar
+```
+
+### Depends On
+Similar to [systemd](https://www.freedesktop.org/wiki/Software/systemd/) this would allow you to specify that a container depends on another container, preventing that container from starting until the container it depends on has also started. This could also be used in shutdown to ensure that the containers which have dependent containers are only terminated after their dependents have all safely shut down.
+```yaml
+containers:
+  - name: myApp
+    dependsOn: mySidecar
+  - name: mySidecar
+```
+This was rejected as the UX was considered to be overly complicated for the use cases that we were trying to solve. It also doesn't solve the problem of Job termination where you want the sidecars to be terminated once the main containers have exited, to do that you would require some additional fields that would further complicate the UX.
+
+### Startup/Shutdown Phases/Numbers
```

Comment:
```suggestion
### Use arbitrary numbers to select the startup/shutdown order
```

Reply:
It wasn't just numbers, we also had the idea of phases which are essentially enums to achieve a similar thing

Reply:
Oh, okay. I think something more descriptive might make sense, though. Not sure this is a good name, but maybe something like this, then?

```suggestion
Predefined startup/shutdown phases (ala "run-level" in sysv init systems)
```